File: test/features/kiosk/kiosk_peripheral_contract_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:larenor/features/kiosk/domain/kiosk_peripheral_contract.dart';

const authority = KioskPeripheralAuthority(
  deviceRevision: 7,
  policyRevision: 11,
  sessionEpoch: 13,
  routeEpoch: 17,
  lifecycleEpoch: 19,
);

Map<String, Object?> provider(
  String id,
  String kind, {
  int revision = 3,
  bool supported = true,
  bool enabledByUser = true,
  String permission = 'granted',
  bool connected = true,
  bool requiresGms = false,
  int maxPayloadBytes = 512,
}) => {
  'providerId': id,
  'kind': kind,
  'revision': revision,
  'supported': supported,
  'enabledByUser': enabledByUser,
  'permission': permission,
  'connected': connected,
  'requiresGms': requiresGms,
  'maxPayloadBytes': maxPayloadBytes,
};

Map<String, Object?> inventoryRaw() => {
  'schemaVersion': 1,
  'inventoryRevision': 5,
  'providers': [
    provider('qr.local_camera', 'qr'),
    provider('nfc.android', 'nfc', permission: 'denied'),
    provider('ble.local', 'ble'),
    provider('ble.play_services', 'ble', requiresGms: true),
    provider('usb.hid', 'usb', connected: false),
    provider(
      'tts.android',
      'tts',
      permission: 'notRequired',
      maxPayloadBytes: 0,
    ),
    provider(
      'print.android',
      'print',
      permission: 'notRequired',
      maxPayloadBytes: 0,
    ),
  ],
};

Map<String, Object?> inputRaw({
  String eventId = '0123456789abcdef0123456789abcdef',
  String providerId = 'qr.local_camera',
  String kind = 'qr',
  int capabilityRevision = 3,
  int sequence = 1,
  int capturedAtElapsedMs = 49000,
  String payload = 'javascript:alert(1)',
}) => {
  'schemaVersion': 1,
  'eventId': eventId,
  'providerId': providerId,
  'kind': kind,
  'capabilityRevision': capabilityRevision,
  'deviceRevision': 7,
  'policyRevision': 11,
  'sessionEpoch': 13,
  'routeEpoch': 17,
  'lifecycleEpoch': 19,
  'sequence': sequence,
  'capturedAtElapsedMs': capturedAtElapsedMs,
  'payload': payload,
};

void main() {
  test('inventory separates opt-in, permission, connection and GMS state', () {
    final inventory = KioskPeripheralInventory.fromChannel(
      inventoryRaw(),
      gmsAvailable: false,
    );

    expect(inventory.providers.map((item) => item.kind).toSet(), {
      KioskPeripheralKind.qr,
      KioskPeripheralKind.nfc,
      KioskPeripheralKind.ble,
      KioskPeripheralKind.usb,
      KioskPeripheralKind.tts,
      KioskPeripheralKind.print,
    });
    expect(
      inventory.provider('qr.local_camera').availability,
      KioskPeripheralAvailability.ready,
    );
    expect(
      inventory.provider('nfc.android').availability,
      KioskPeripheralAvailability.permissionDenied,
    );
    expect(
      inventory.provider('ble.play_services').availability,
      KioskPeripheralAvailability.gmsUnavailable,
    );
    expect(
      inventory.provider('usb.hid').availability,
      KioskPeripheralAvailability.disconnected,
    );

    final disabled = inventoryRaw();
    (disabled['providers']! as List<Object?>)[0] = provider(
      'qr.local_camera',
      'qr',
      enabledByUser: false,
    );
    expect(
      KioskPeripheralInventory.fromChannel(
        disabled,
        gmsAvailable: false,
      ).provider('qr.local_camera').availability,
      KioskPeripheralAvailability.disabled,
    );
  });

  test('input stays untrusted review-only data and rejects unsafe scope', () {
    final inventory = KioskPeripheralInventory.fromChannel(
      inventoryRaw(),
      gmsAvailable: false,
    );
    final gate = KioskPeripheralInputGate();
    final input = gate.accept(
      inputRaw(),
      inventory: inventory,
      expectedAuthority: authority,
      isCurrent: () => true,
      nowElapsedMs: 50000,
    );

    expect(input.payload, 'javascript:alert(1)');
    expect(input.disposition, KioskPeripheralDisposition.reviewOnly);
    expect(input.canExecuteCommand, isFalse);
    expect(input.canInjectJavaScript, isFalse);
    expect(jsonEncode(input.toAuditJson()), isNot(contains(input.payload)));

    for (final invalid in [
      inputRaw(providerId: 'nfc.android', kind: 'nfc'),
      inputRaw(providerId: 'usb.hid', kind: 'usb'),
      inputRaw(providerId: 'tts.android', kind: 'tts'),
      inputRaw(capabilityRevision: 4),
      {...inputRaw(), 'routeEpoch': 18},
      {...inputRaw(), 'extra': 'execute'},
      inputRaw(payload: 'x' * 513),
    ]) {
      expect(
        () => KioskPeripheralInputGate().accept(
          invalid,
          inventory: inventory,
          expectedAuthority: authority,
          isCurrent: () => true,
          nowElapsedMs: 50000,
        ),
        throwsA(isA<FormatException>()),
      );
    }
  });

  test('duplicate, replay, stale and revoked input fail closed', () {
    final inventory = KioskPeripheralInventory.fromChannel(
      inventoryRaw(),
      gmsAvailable: false,
    );
    final gate = KioskPeripheralInputGate();
    gate.accept(
      inputRaw(sequence: 9),
      inventory: inventory,
      expectedAuthority: authority,
      isCurrent: () => true,
      nowElapsedMs: 50000,
    );

    for (final replay in [
      inputRaw(sequence: 9),
      inputRaw(eventId: '1123456789abcdef0123456789abcdef', sequence: 8),
      inputRaw(
        eventId: '2123456789abcdef0123456789abcdef',
        sequence: 10,
        capturedAtElapsedMs: 1000,
      ),
    ]) {
      expect(
        () => gate.accept(
          replay,
          inventory: inventory,
          expectedAuthority: authority,
          isCurrent: () => true,
          nowElapsedMs: 50000,
        ),
        throwsA(isA<FormatException>()),
      );
    }
    expect(
      () => gate.accept(
        inputRaw(eventId: '3123456789abcdef0123456789abcdef', sequence: 10),
        inventory: inventory,
        expectedAuthority: authority,
        isCurrent: () => false,
        nowElapsedMs: 50000,
      ),
      throwsA(isA<FormatException>()),
    );
  });
}

File: android/app/src/test/kotlin/com/ersingundem/larenor/kiosk/KioskPeripheralContractTest.kt
package com.ersingundem.larenor.kiosk

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.fail
import org.junit.Test

class KioskPeripheralContractTest {
    private val authority = PeripheralAuthority(7, 11, 13, 17, 19)

    private fun capabilities() = KioskPeripheralInventory(
        inventoryRevision = 5,
        gmsAvailable = false,
        providers = listOf(
            PeripheralCapability("qr.local_camera", PeripheralKind.qr, 3, true, true, PeripheralPermission.granted, true, false, 512),
            PeripheralCapability("nfc.android", PeripheralKind.nfc, 3, true, true, PeripheralPermission.denied, true, false, 512),
            PeripheralCapability("ble.local", PeripheralKind.ble, 3, true, true, PeripheralPermission.granted, true, false, 512),
            PeripheralCapability("ble.play_services", PeripheralKind.ble, 3, true, true, PeripheralPermission.granted, true, true, 512),
            PeripheralCapability("usb.hid", PeripheralKind.usb, 3, true, true, PeripheralPermission.granted, false, false, 512),
            PeripheralCapability("tts.android", PeripheralKind.tts, 3, true, true, PeripheralPermission.notRequired, true, false, 0),
            PeripheralCapability("print.android", PeripheralKind.print, 3, true, true, PeripheralPermission.notRequired, true, false, 0),
        ),
    )

    private fun event(id: String = "0123456789abcdef0123456789abcdef", sequence: Int = 1) = mapOf(
        "schemaVersion" to 1,
        "eventId" to id,
        "providerId" to "qr.local_camera",
        "kind" to "qr",
        "capabilityRevision" to 3,
        "deviceRevision" to 7,
        "policyRevision" to 11,
        "sessionEpoch" to 13,
        "routeEpoch" to 17,
        "lifecycleEpoch" to 19,
        "sequence" to sequence,
        "capturedAtElapsedMs" to 49_000L,
        "payload" to "javascript:alert(1)",
    )

    private fun fails(action: () -> Unit) {
        try { action(); fail("expected fail-closed rejection") } catch (_: PeripheralFailure) {}
    }

    @Test fun inventorySeparatesOptInPermissionConnectionAndGms() {
        val inventory = capabilities()
        assertEquals(PeripheralAvailability.ready, inventory.provider("qr.local_camera").availability(false))
        assertEquals(PeripheralAvailability.permissionDenied, inventory.provider("nfc.android").availability(false))
        assertEquals(PeripheralAvailability.gmsUnavailable, inventory.provider("ble.play_services").availability(false))
        assertEquals(PeripheralAvailability.disconnected, inventory.provider("usb.hid").availability(false))
        assertEquals(6, inventory.providers.map { it.kind }.toSet().size)
        assertEquals(
            PeripheralAvailability.disabled,
            inventory.provider("qr.local_camera").copyForTest(enabledByUser = false).availability(false),
        )
    }

    @Test fun inputIsBoundedReviewOnlyAndNeverExecutes() {
        val accepted = KioskPeripheralInputGate().accept(event(), capabilities(), authority, true, 50_000L)
        assertEquals("javascript:alert(1)", accepted.payload)
        assertEquals(PeripheralDisposition.reviewOnly, accepted.disposition)
        assertFalse(accepted.canExecuteCommand)
        assertFalse(accepted.canInjectJavaScript)

        fails { KioskPeripheralInputGate().accept(event() + ("providerId" to "nfc.android") + ("kind" to "nfc"), capabilities(), authority, true, 50_000L) }
        fails { KioskPeripheralInputGate().accept(event() + ("providerId" to "tts.android") + ("kind" to "tts"), capabilities(), authority, true, 50_000L) }
        fails { KioskPeripheralInputGate().accept(event() + ("payload" to "x".repeat(513)), capabilities(), authority, true, 50_000L) }
        fails { KioskPeripheralInputGate().accept(event() + ("sequence" to 1.0), capabilities(), authority, true, 50_000L) }
    }

    @Test fun duplicateReplayStaleAndRevokedEventsAreRejected() {
        val gate = KioskPeripheralInputGate()
        gate.accept(event(sequence = 9), capabilities(), authority, true, 50_000L)
        fails { gate.accept(event(sequence = 9), capabilities(), authority, true, 50_000L) }
        fails { gate.accept(event("1123456789abcdef0123456789abcdef", 8), capabilities(), authority, true, 50_000L) }
        fails { gate.accept(event("2123456789abcdef0123456789abcdef", 10) + ("capturedAtElapsedMs" to 1_000L), capabilities(), authority, true, 50_000L) }
        fails { gate.accept(event("3123456789abcdef0123456789abcdef", 10), capabilities(), authority, false, 50_000L) }
    }

    private fun PeripheralCapability.copyForTest(enabledByUser: Boolean) = PeripheralCapability(
        providerId, kind, revision, supported, enabledByUser, permission, connected, requiresGms, maxPayloadBytes,
    )
}
